File: pa4/List.c
//Kody Fong 1519213
//kbfong
//PA 2
//List ADT but in C. Follows the structure of List.java but implemented in C

#include <stdio.h>
#include <stdlib.h>
#include "List.h"



typedef struct NodeObj{
    int data;  
    struct NodeObj* prev;  
    struct NodeObj* next;  
} NodeObj;

typedef NodeObj* Node;


typedef struct ListObj {
    Node front;  
    Node back;  
    Node cursor;
    int index;  
    int length;  
} ListObj;


//creates a new node object and initializes the data, prev, and next
Node newNode(int value) {
    Node temp = malloc(sizeof(NodeObj));
    temp->data = value;
    temp->prev = NULL;
    temp->next = NULL;
    return(temp);
}

//frees the memory from the heap for the node then sets it to null
void freeNode(Node* pN)
{
    if( pN!=NULL && *pN!=NULL )
    {
        free(*pN);
        *pN = NULL;
    }
}


//creates new empty list
List newList(void) {
  List newList; 
  newList = malloc(sizeof(struct ListObj)); //allocates memory
  newList->front = NULL;
  newList->back = NULL;
  newList->cursor = NULL;
  newList->length = 0;
  newList->index = -1;
  return newList;
}

//frees the memory from the heap for the list and sets it to null
void freeList(List* pL)
{
    if(pL != NULL && *pL != NULL)
    {
        
        while ( length(*pL) > 0 ) 
        {
            deleteFront(*pL);  // delete front node in List
        }
        free(*pL); //frees the memory from the list.
        *pL = NULL; //sets pL to null.
    }
    
}


// Returns the number of elements in this List.
int length(List L)
{
    return (L->length); //returns the length


}

// If cursor is defined, returns the index of the cursor element, otherwise returns -1.
int index(List L)
{
    if(L->cursor == NULL)
    {
        return -1;
    }
    return L->index; //returns the index
}

// Returns front element. Pre: length()>0
int front(List L)
{
   if(L->length == 0)
   {
    printf("List Error: Calling front() on empty List");
    exit(1);
   }
   return (L->front->data); //returns the front 
}

// Returns back element. Pre: length()>0
int back(List L)
{
   if(L->length == 0)
   {
    printf("List Error: Calling back() on empty List");
    exit(1);
   }
   return (L->back->data); //returns the back
}

// Returns cursor element. Pre: length()>0, index()>=0
int get(List L)
{
    if (L->length ==0)
    {
        printf("List Error: Calling get() on empty List");
        exit(1);
    }
    if(L->cursor == NULL)
    {
        printf("List Error: Calling get() on cursor that is off the list");
        exit(1);
    }
    return (L->cursor->data); //returns the data at the cursor
}

//// Returns 1(true) if and only if  List A and B are the same integer sequence. Else return 0(false).
int equals(List A, List B)
{
    int isEqual = 1; //stores the return value after checking if the lists are equal
    Node aFront = A->front; //sets a temp node for the front node of List A
    Node bFront = B->front; //sets a temp node for the front node of List B
    if(A->length == B-> length) 
    {
        while( isEqual == 1 && aFront != NULL && bFront != NULL) //while the values are equal and the nodes aren't null
        {
            if (aFront->data == bFront->data) //if the data is the same go to the next node and check 
            {
                aFront = aFront->next; 
                bFront = bFront->next;
            }
            else
            {
                isEqual = 0; //if they're not equal then return 0 (false) and the loop ends.
            }
        }
        return isEqual;
    }
    else //the Lists aren't the same length therefore they cannot be equivalent so return 0.
    {
        isEqual = 0;
        return isEqual;
    }
}

// Manipulation procedures ----------------------------------------------------

// Resets this List to its original empty state.
void clear(List L)
{
   while (L->length > 0) {
        deleteFront(L);  // delete front node in List
    }
   L->front = NULL;
   L->back = NULL;
   L->cursor = NULL;
   L->length = 0;
   L->index = -1;
}

// If List is non-empty, places the cursor under the front element, otherwise does nothing.
void moveFront(List L)
{
   if(L->length !=0)
   {
    L->cursor = L->front;
    L->index = 0; //cursor will be at the front
   }
}

// If List is non-empty, places the cursor under the back element, otherwise does nothing.
void moveBack(List L)
{
    if(L->length !=0)
    {
        L->cursor = L->back;
        L->index = L->length - 1; //cursor will be at the back
    }
}

// If cursor is defined and not at front, moves cursor one step toward front of this List, if cursor is defined and at front, cursor becomes undefined, if cursor is undefined does nothing.
void movePrev(List L)
{
    if(L->cursor != NULL)
    {
        if(L->cursor == L->front)
        {
            L->cursor = NULL;
            L->index = -1;
        }
        else
        {
            L->cursor = L->cursor->prev; //moves cursor towards front of the list
            L->index--; //decrements the index
        }
    }
}

void moveNext(List L)
{
    if(L->cursor != NULL)
    {
        if(L->cursor == L->back)
        {
            L->cursor = NULL;
            L->index = -1;

        }
        else
        {
            L->cursor = L->cursor->next; //moves cursor towards the back of the list
            L->index++; //increments the index
        }
    }
}

// Insert new element into this List. If List is non-empty, insertion takes place before front element.
void prepend(List L, int data) 
{
    Node temp = newNode(data);
    if (L->length != 0) 
    {
        L->front->prev = temp; //the front of the list's previous becomes the temp node making the temp node first in the list
        temp->next = L->front; //the temp nodes next will be the original front of the list
        L->front = temp;       //the new front of the list becomes the temp node
        L->index++;
        L->length++;
    }
    else //the list is empty so just add to the list
    {  
        L->front = L->back = L->cursor = temp; 
        L->index = 0;
        L->length++;
    }
 
}

// Insert new element into this List. If List is non-empty, insertion takes place after back element.
void append(List L, int data) 
{
    Node temp = newNode(data);
    if (L->length != 0) 
    {
        temp->prev = L->back; //sets the temp node's previous value to the original back of the list
        L->back->next = temp; //the back of the list's next value will be the temp node
        L->back = temp;       //the new back of the list becomes the temp node
        L->length++;
        
    } 
    else 
    {  // the list is empty so just add to the list
        L->front = L->back = L->cursor = temp;
        L->index = 0;
        L->length++;
    }
    
}

// Insert new element before cursor. Pre: length()>0, index()>=0
void insertBefore(List L, int data)
{
    
    if(L->length == 0)
    {
        printf("List Error: Cannot call insertbefore() on empty list");
        exit(1);
    }

    if(L->cursor == NULL)
    {
        printf("List Error: Cannot call insertbefore() on cursor that is off the list");
        exit(1);
    }

    if(L->cursor == L->front)
    {
        prepend(L, data); //if the cursor is at the front of the list, you can just call prepend since you will be inserting the element before the front element 
    }
    else
    {
        Node temp = newNode(data);
        temp->prev = L->cursor->prev; //the temp nodes previous will become the cursor's previous
        temp->next = L->cursor;       //the temp nodes next will be the cursor
        L->cursor->prev->next = temp;  //cursors previous next element will be temp
        L->cursor->prev = temp;        //cursors previous will be temp
        L->length++;
        L->index++;
    }
}   

// Inserts new element after cursor. Pre: length()>0, index()>=0
void insertAfter(List L, int data)
{
    
    if(L->length == 0)
    {
        printf("List Error: Cannot call insertAfter() on empty list");
        exit(1);
    }

    if(L->cursor == NULL)
    {
        printf("List Error: Cannot call insertAfter() on cursor that is off the list");
        exit(1);
    }

    if(L->cursor == L-> back)
    {
        append(L, data); //if the cursor is at the back of the list then you can just all append since you will be inserting the element into the back element
    }

    else
    {
        Node temp = newNode(data); 
        temp->next = L->cursor->next; //the new node temp's next will be the cursor's next
        temp->prev = L->cursor;       //the new node temp's previous will become the cursor
        L->cursor->next->prev = temp; //cursor's next previous element becomes temp
        L->cursor->next = temp;       //cursor's next becomes temp
        L->length++;
    }

}

// Deletes the front element. Pre: length()>0
void deleteFront(List L) 
{
    if(L->length == 0 )
    {
        printf("List Error: Cannot call deleteFront() on empty list");
        exit(1);
    }
    if (L->length== 1) 
    {
        Node temp = L->front; //temp node becomes equal to the front
        L->front = L->cursor = L->back = NULL;
        L->index = -1;
        L->length = 0;
        freeNode(&temp);
    } 
    else 
    {
        if (L->cursor == L->front) 
        {
             L->cursor = NULL;
             L->index = -1;
        } 
       
    Node temp = L->front; //creates a node and sets it equal to the front of the list
    L->front = L->front->next; //the front of the list becomes the next value within the list
    L->front->prev = NULL; //the previous element in the front becomes null , the temp node becomes null 
    L->length--;
    L->index--;
    freeNode(&temp); //frees the memory from the temp nodes
        
    }
}

// Deletes the back element. Pre: length()>0
void deleteBack(List L) 
{
    if(L->length ==0)
    {
        printf("List Error: Cannot call deleteBack() on empty list");
        exit(1);
    }
    if (L->length == 1) 
    {
         Node temp = L->back; //temp nodes becomes equal to the back
         L->front = L->cursor = L->back = NULL;
         L->index= -1;
         L->length = 0;
         freeNode(&temp);
    } 
    else 
    {
        if (L->cursor == L->back)
         {
            L->cursor = NULL;
            L->index = -1;
         }
            Node temp = L->back; //creates  a node and sets it equal to the back of the list
            L->back = L->back->prev; //the new back of the list becomes the previous value within the list
            L->back->next = NULL; //the next element in the back of the list becomes null, the temp node becomes null
            L->length--;
            freeNode(&temp);
        }
    
}

// Deletes cursor element, making cursor undefined. Pre: length()>0, index()>=0
void delete(List L) 
{
    if(L->length == 0 )
    {
        printf("List Error: Cannot call delete() on empty list");
        exit(1);
    }

    if(L->cursor == NULL)
    {
        printf("List Error: Cannot call delete() on cursor that is off the list");
        exit(1);
    }
    
    if(L->cursor == L->front)
    {
        deleteFront(L); //deletes the front if the cursor is at the front
    }

    else if(L->cursor == L->back)
    {
        deleteBack(L); //deletes the back if the cursor is at the back
    }
    else
    {
        Node temp = L->cursor; //sets temp node to the cursor
        L->cursor->prev->next = L->cursor->next; //the cursors previours next becomes the cursors next
        L->cursor->next->prev = L->cursor->prev; //the cursors next's previous becomes the cursors previous
        L->cursor = NULL;
        L->index = -1;
        L->length--;
        freeNode(&temp);

    }
    
}


// Prints List
void printList(FILE* out, List L)
{
    Node temp = L->front; //sets a temp node equal to the front of the list
    int i = 0; //counter
    while(temp != NULL)   //while the temp node isn't null print the values within the list
    {
        if(i !=0){
         fprintf(out, " %d", temp->data); //print out the value of the node's data within the list
        }
      else{

        fprintf(out, "%d", temp->data);
        

      }
      temp = temp->next;  //move on to the next value within the list
      i++;
        
    }     
}

// Returns a new List representing the same integer sequence as this List. The cursor in the new list is undefined, regardless of the state of the cursor in this List. This List is unchanged.
List copyList(List L)
{
    List n = newList(); //creates a new list
    Node temp = L->front; //creates a node and sets it equal to the front of the list
    while(temp!= NULL) //while the temp node isn't null copy the values into the list
    {
        append(n,temp->data); //appends the data from the current node onto the list
        temp = temp->next; //move to the next value within the list
    }
    return n; //returns the new list after being copied
}

File: pa3/List.java
/*
Kody Fong (1519213)
kbfong
PA3
List.java is a program that implements all the required functions for a linked list that was from previous pa's. This time, the list utilizes Objects 


*/


public class List
{
	private class Node
	{
		Object data;
		Node next;
		Node prev;

		Node(Object nData) 
		{
			data = nData;
			next = null;
			prev = null;
		}

		public String toString() //overriding of the toString method 
		{
			return String.valueOf(data);
		}
		public boolean equals(Object x) //overrides equals method. Was failing the NonEmpty equals test because I forgot to implement this.
									    //Thanks to Edward Chan who answered my question on Piazza https://piazza.com/class/jtrpcey11c221r?cid=532
		{
			Node temp = (Node) x;
			boolean isEqual = false;
			if(data.equals(temp))
			{
				isEqual = true;
			}
			return isEqual;

		}
	}

	private Node front;
	private Node back;
	private Node cursor;
	private int index;
	private int length;

	List()
	{
		front = null;
		back = null;
		cursor = null;
		length = 0;
		index = -1;
	}
	// Returns the number of elements in this List.
	int length()
	{
		return length;
	}
	// If cursor is defined, returns the index of the cursor element, otherwise returns -1.
	int index(){

		
		if (cursor == null) 
		{
			return -1;
		}
		return index;
	}
	// Returns front element. Pre: length()>0
	Object front()
	{
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call front() on empty List");
		}
		return front.data;
	}
	// Returns back element. Pre: length()>0
	Object back()
	{
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call back() on empty List");
		}
		return back.data;
	}
	// Returns cursor element. Pre: length()>0, index()>=0
	Object get()
	{
		if (length == 0 ) 
		{	
			throw new RuntimeException("List Error: cannot call get() on empty List");
		}
		if (cursor == null	) 
		{
			throw new RuntimeException("List Error: cannot call get() on cursor that is off the list");
			
		}
		else
		{
			return cursor.data;
		}
	}
	//// Returns true if and only if this List and L are the same integer sequence
	public boolean equals(Object L)
	{
		List tempList = (List) L; //creates a List object 

		boolean isEqual = true;
		Node thisTempFront = front;  
		Node thatTempFront = tempList.front;        
		if (length == tempList.length) //check if the List and L are at least the same length
		 {
			while( isEqual == true && thisTempFront != null && thatTempFront != null) 
			{
				if (thisTempFront.data.equals(thatTempFront.data)) //if the data within each node is equivalent move on to the next node to check
				{

					thatTempFront = thatTempFront.next;
					thisTempFront = thisTempFront.next;

				}
				else
				{
					isEqual = false; //if the data is not the same within each node then the integer sequence is not the same and the loop ends
				}
			}
			return isEqual;
		}
		else  //the List and L are not the same length, therefore they cannot be the same integer sequence
		{
			isEqual = false;
			return isEqual;
		}
	}
	//Manipulation procedures
	// Resets this List to its original empty state.
	void clear()
	{
		front = null;
		back = null;
		cursor = null;
		length = 0;
		index = -1;
	}
	// If List is non-empty, places the cursor under the front element, otherwise does nothing.
	void moveFront()
	{
		if (length != 0)
		 {
			cursor = front;
			index = 0; //cursor will be at the front
			

		}
	}
	// If List is non-empty, places the cursor under the back element, otherwise does nothing.
	void moveBack()
	{
		if (length != 0) 
		{
			cursor = back; 
			index = length - 1; //cursor will be at the back
			
		}
	}
	// If cursor is defined and not at front, moves cursor one step toward front of this List, if cursor is defined and at front, cursor becomes undefined, if cursor is undefined does nothing.
	void movePrev()
	{
		if (cursor != null) 
		{
			if (cursor == front) 
			{
				cursor = null;
                index = -1; //added this because i kept failing the test cases on the grading script
			}
			else
			{
				cursor = cursor.prev; //moves cursor toward front of the list
                index--; //added this because i kept failing the test cases on the grading script
			}
		}
	}
	// If cursor is defined and not at back, moves cursor one step toward back of this List, if cursor is defined and at back, cursor becomes undefined, if cursor is undefined does nothing.
	void moveNext()
	{
		if (cursor != null) 
		{
			if (cursor == back) 
			{
				cursor = null;
                index = -1; //added cause i was failing the test scripts and forgot to manage the indices 
			}
			else
			{
				cursor = cursor.next; //moves cursor toward back of the list
                index++; //added cause i was failing to the test scripts and forgot to manage the indicies
			}
		}
	}
	// Insert new element into this List. If List is non-empty, insertion takes place before front element.
	void prepend(Object data)
	{
		Node temp = new Node(data);
		if (front == null) 
		{
			front = back = temp; // if the list is empty the node will be the first node in the list making it the front and back node
			length++;
                        index = 0; //added because i failed the test scripts because i wasn't managing the indicies
		}
		else
		{
			front.prev=temp; //the temp node is inserted before the front
			temp.next=front;     //the node after temp will be the front because temp was inserted before the front
			front = temp;        //the front is now the temp node
			length++;
            index++;//added because i failed the test scripts because i wasn't managing the indicies
		}
	}

	// Insert new element into this List. If List is non-empty, insertion takes place after back element.
	void append(Object data)
	{
		Node temp = new Node(data);
		if (back == null) 
		{
			front = back = temp; //if the list is empty, then the new node temp becomes the front and back of the list as its the only node within the list
			length++;
		}
		else
		{
			back.next = temp; //the next node after the back node becomes the new node temp since it was inserted into the back of the list
			temp.prev = back; //the new node temp's previous node becomes the back since temp is now the new back of the list
			back = temp; //the new back of the list now becomes the new node temp
			length++;
		}
	}
	// Insert new element before cursor. Pre: length()>0, index()>=0
	void insertBefore(Object data)
	{
		Node temp = new Node(data);
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call insertBefore() on empty List");
		}
		if (cursor == null) 
		{
			throw new RuntimeException("List Error: cannot call insertBefore() on cursor that is off the list");
		}
		if (cursor == front) 
		{
			prepend(data);  //if the cursor is at the front of the list, you can just call prepend since you will be inserting the element before the front element 
		}
		else
		{
			
			temp.prev = cursor.prev; //the new node temp's previous becomes the cursors previous	
            temp.next = cursor;      //the new node temp's next becomes the cursor
            cursor.prev.next = temp; //cursor's previous next element becomes temp
            cursor.prev = temp;      //cursor's previous becomes temp
            length++;
            index++;
		}
	}
	// Inserts new element after cursor. Pre: length()>0, index()>=0
	void insertAfter(Object data)
	{
		Node temp = new Node(data);
		if (length==0) 
		{
			throw new RuntimeException("List Error: cannot call insertAfter() on empty List");
		}
		if (cursor == null) 
		{
			throw new RuntimeException("List Error: cannot call insertAfter() on cursor that is off the list");
		}
		if (cursor == back) 
		{
			append(data); //if the cursor is at the back of the list then you can just all append since you will be inserting the element into the back element
		}
		else
		{
			
			temp.next = cursor.next; //the new node temp's next will be the cursor's next
			temp.prev = cursor;      //the new node temp's previous will become the cursor
			cursor.next.prev = temp;  //cursor's next previous element becomes temp
			cursor.next = temp;      //cursor's next becomes temp
			length++;
		}
	}
	// Deletes the front element. Pre: length()>0
	void deleteFront()
	{
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call deleteFront() on empty List");
		}
		if (cursor == front) 
		{
			cursor = null;
			index = -1; //added because i was failing the test scripts because i wasn't managing the indicies
		}
        
		front = front.next; //front element becomes the next element in the list
		length--;
        index --; //added because i was failing the test scripts because i wasn't managing the indicies
		
	}

	// Deletes the back element. Pre: length()>0
	void deleteBack()
	{
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call deleteBack() on empty List");
		}
		if (cursor == back) 
		{
			cursor = null;
			index = -1;
		}
        
		back = back.prev; //the back element becomes the previous element in the list
		length--;

	}

	// Deletes cursor element, making cursor undefined. Pre: length()>0, index()>=0
	void delete()
	{
		if (length == 0) 
		{
			throw new RuntimeException("List Error: cannot call delete() on empty List");
		}
		if (cursor == null) 
		{
			throw new RuntimeException("List Error: cannot call delete() on cursor that is off the list");
		}
		if (cursor == front) 
		{
			deleteFront(); //deletes the front if the cursor is at the front

		}
		else if(cursor == back)
		{
			deleteBack(); //deletes the back if the cursor is at the back
		}
		else
		{
			cursor.prev.next=cursor.next;  
			cursor.next.prev=cursor.prev;
			cursor = null; 
			index = -1;
			length--;
		}
		
	}

	// Overrides Object's toString method. Returns a String representation of this List consisting of a space separated sequence of integers, with front on left.
	public String toString()
	{
		String stringList = "";
		Node temp = front;
		while(temp != null)
		{
			stringList += String.valueOf(temp.data) + " "; //places the data from the Node into the empty string
			temp = temp.next;              //moevs onto the next Node in the list
		}
		return stringList;
	}
	
	// Returns a new List representing the same integer sequence as this List. The cursor in the new list is undefined, regardless of the state of the cursor in this List. This List is unchanged.
	//List copy()
	//{
	//	List newList = new List();
//		Node temp = front;
//		while (temp!=null) 
//		{
//			newList.append(temp.data); //calls append on the new list
//			temp = temp.next;
//		}
//		newList.cursor = null; //undefined cursor
//		return newList;
//	}


}

